pkg/service/auth_context.go

```diff
@@ -23,44 +27,125 @@ type AuthContext struct {
 	Authorization map[*config.AuthorizationConfig]interface{}
 }
 
-// Evaluate evaluates all steps of the auth pipeline (identity → metadata → policy enforcement)
-func (self *AuthContext) Evaluate() error {
-	self.Identity = make(map[*config.IdentityConfig]interface{})
-	self.Metadata = make(map[*config.MetadataConfig]interface{})
-	self.Authorization = make(map[*config.AuthorizationConfig]interface{})
-
-	// identity (authentication)
-	identityConfigs := self.API.IdentityConfigs
-	for i := range identityConfigs {
-		c := identityConfigs[i]
-		if ret, err := c.Call(self); err != nil {
+// AuthObjectConfig provides an interface for APIConfig objects that implements a Call method
+type AuthObjectConfig interface {
+	Call(ctx common.AuthContext) (interface{}, error)
+}
+
+type configCallback = func(config AuthObjectConfig, obj interface{})
+
+// NewAuthContext creates an AuthContext instance
+func NewAuthContext(parentCtx context.Context, req *auth.CheckRequest, apiConfig config.APIConfig) AuthContext {
+
+	return AuthContext{
+		ParentContext: &parentCtx,
+		Request:       req,
+		API:           &apiConfig,
+		Identity:      make(map[*config.IdentityConfig]interface{}),
+		Metadata:      make(map[*config.MetadataConfig]interface{}),
+		Authorization: make(map[*config.AuthorizationConfig]interface{}),
+	}
+
+}
+
+func (authContext *AuthContext) getAuthObject(ctx context.Context, objConfig AuthObjectConfig, cb configCallback) error {
+	log := ctrl.Log.WithName("Authorino").WithName("AuthContext")
+	select {
+	case <-ctx.Done():
+		log.Info("Context cancelled objConfig %v terminting", objConfig)
+		return nil
+	default:
+		if authObj, err := objConfig.Call(authContext); err != nil {
+			log.Error(err, "Invalid auth object config")
 			return err
 		} else {
-			self.Identity[&c] = ret
-			break
+			cb(objConfig, authObj)
+			return nil
 		}
 	}
+}
+
+func (authContext *AuthContext) getAuthObjects(configs []AuthObjectConfig, cb configCallback) error {
+	log := ctrl.Log.WithName("Authorino").WithName("AuthContext")
+	errGroup, ctx := errgroup.WithContext(context.Background())
+
+	for _, config := range configs {
+		objConfig := config
+		errGroup.Go(func() error {
+			return authContext.getAuthObject(ctx, objConfig, cb)
+		})
+	}
+	if err := errGroup.Wait(); err != nil {
+		return err
+	} else {
+		log.Info("Successfully fetched all auth objects.", configs)
+		return nil
+	}
+}
+
+// GetIDObject gets an Identity auth object given an Identity config.
+func (authContext *AuthContext) GetIDObject() error {
```

Comment:
```suggestion
func (authContext *AuthContext) GetIdentityObject() error {
```

Reply:
After reading again, I have another suggestion for the name of this function (and analogously extensible to the other 2 below):

```suggestion
func (authContext *AuthContext) EvaluateIdentity() error {
```

...because this is about splitting `Evaluate()` into the part that is specific about identity, right? `GetWhateverObject` doesn't say anything to me really.